internal/controllers/machinedeployment/machinedeployment_rolling.go

```diff
@@ -160,204 +169,177 @@ func (p *rolloutPlanner) reconcileNewMachineSet(ctx context.Context) error {
+	// This is required because whenever we are reducing the number of replicas, this operation could further impact availability e.g.
+	// - in case of regular rollout, there is no certainty about which machine is going to be deleted (and if this machine is currently available or not):
+	// 	 - e.g. MS controller is going to delete first machines with deletion annotation; also MS controller has a slight different notion of unavailable as of now.
+	// - in case of in-place rollout, in-place upgrade are always assumed as impacting availability (they can always fail).
+	totPendingScaleDown := int32(0)
+	for _, ms := range allMSs {
+		scaleIntent := ptr.Deref(ms.Spec.Replicas, 0)
+		if v, ok := p.scaleIntents[ms.Name]; ok {
+			scaleIntent = min(scaleIntent, v)
+		}
+
+		// NOTE: we are counting only pending scale down from the current status.replicas (so scale down of actual machines).
+		if scaleIntent < ptr.Deref(ms.Status.Replicas, 0) {
+			totPendingScaleDown += max(ptr.Deref(ms.Status.Replicas, 0)-scaleIntent, 0)
+		}
 	}
 
-	// Clean up unhealthy replicas first, otherwise unhealthy replicas will block deployment
-	// and cause timeout. See https://github.com/kubernetes/kubernetes/issues/16737
-	cleanupCount, err := p.cleanupUnhealthyReplicas(ctx, maxScaledDown)
-	if err != nil {
-		return err
+	// Compute the total number of replicas that can be scaled down.
+	// Exit immediately if there is no room for scaling down.
+	totalScaleDownCount := max(totReplicas-totPendingScaleDown-minAvailable, 0)
+	if totalScaleDownCount <= 0 {
+		return nil
 	}
 
-	log.V(4).Info("Cleaned up unhealthy replicas from old MachineSets", "count", cleanupCount)
+	sort.Sort(mdutil.MachineSetsByCreationTimestamp(p.oldMSs))
 
-	// Scale down old MachineSets, need check maxUnavailable to ensure we can scale down
-	scaledDownCount, err := p.scaleDownOldMachineSetsForRollingUpdate(ctx)
-	if err != nil {
-		return err
-	}
+	// Scale down only unavailable replicas / up to residual totalScaleDownCount.
+	// NOTE: we are scaling up unavailable machines first in order to increase chances for the rollout to progress;
+	// however, the MS controller might have different opinion on which machines to scale down.
+	// As a consequence, the scale down operation must continuously assess if reducing the number of replicas
+	// for an older MS could further impact availability under the assumption than any scale down could further impact availability (same as above).
+	// if reducing the number of replicase might lead to breaching minAvailable, scale down extent must be limited accordingly.
+	totalScaleDownCount, totAvailableReplicas = p.scaleDownOldMSs(ctx, totalScaleDownCount, totAvailableReplicas, minAvailable, false)
+
+	// Then scale down old MS up to zero replicas / up to residual totalScaleDownCount.
+	// NOTE: also in this case, continuously assess if reducing the number of replicase could further impact availability,
```

Comment:
Also having trouble parsing this sentence too. I think you're saying we will scale down the MS decrementing by `totalScaleDownCount` without going below 0?

```suggestion
	// Then scale down old MS by totalScaleDownCount to get to zero.
	// NOTE: also in this case, continuously assess if reducing the number of replicas could further impact availability,
```